### PR TITLE
Upgrade the parent lookup logic

### PR DESCRIPTION
**Pull request by AgeManning**
## Issue Addressed
#822 

## Proposed Changes

Updates the logic for how unknown block lookups function. This update has stricter rules on parent-block lookup chains. Chains will fail if any block in the chain cannot be processed. Peers will be downvoted if they send us a block that cannot be processed. 

This also randomises peers for block lookups.

### CODE REVIEW

**Comment by AgeManning** [comment]
This is ready for review, although I have not done any rigorous test on the logic. It would be nice to test this out on a testnet / localnet before merging. 

Potentially the syncing simulation may be sufficient

File: beacon_node/network/src/sync/manager.rs [hunk 1]
```diff
@@ -631,12 +725,13 @@ impl<T: BeaconChainTypes> SyncManager<T> {
             block_roots: vec![parent_hash],
         };
         // select a random fully synced peer to attempt to download the parent block
-        let peer_id = self.full_peers.iter().next().expect("List is not empty");
+        let available_peers = self.full_peers.iter().collect::<Vec<_>>();
+        let peer_id = (**available_peers
+            .choose(&mut rand::thread_rng())
+            .expect("List is not empty"))
```
Comment (paulhauner) [inline beacon_node/network/src/sync/manager.rs]: Perhaps you could avoid this `expect` by removing the `self.full_peers.is_empty()` check on L704 and doing this instead:

```rust
let peer_id = if let Some(peer_id) = **available_peers.choose(&mut rand::thread_rng()) {
  peer_id.clone()
} else {
  return;
}
```

Comment (paulhauner) [reply beacon_node/network/src/sync/manager.rs]: There's also an `expect` on L722 that I assume would be more helpful if it raised a `CRIT`.

**Comment by AgeManning** [comment]
This potentially needs at add extra cases in block processing

**Pull request merged**